events/events.go

```diff
@@ -31,9 +35,83 @@ type EventListener interface {
 	HandleEvent(event Event)
 }
 
+// EventEmitter describes an interface that can be used to emit events
+type EventEmitter interface {
+	EmitEvent(Event)
+}
+
 // EventRegistrar is an object that you can register EventListeners on.
 type EventRegistrar interface {
 	RegisterListener(EventListener)
+	DeregisterListener(EventListener)
+}
+
+type baseEventRegistrar struct {
+	lock      sync.RWMutex
+	listeners []EventListener
+}
+
+// RegisterListener adds a listener to the Event Registar. Events emitted on this
+// registar will be invoked on the listener
+func (a *baseEventRegistrar) RegisterListener(l EventListener) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// by making a copy the backing array will never be changed after its creation.
+	// this allowes to copy the slice while locked but iterate while not locked
+	// preventing deadlocks when listeners are added/removed in the handler of a
+	// listener
+	cpy := append([]EventListener(nil), a.listeners...)
```

Comment:
If you want to optimize the allocations here it might be better to initialize `cpy` as a new slice with a capacity of `len(a.listeners)+1` and append `a.listeners` and `l` to that one. Would probably save you a re-size/re-allocate on the internal array of the slice.